### DISCUSSION TITLE
Is it possible to plot signals differently?

### DISCUSSION THREAD

**Discussion by Spooyler** [hyperspy/exspy#74] [Q&A] [answered]
As I know the only way to handle signal plotting, is through self.plot(). But my problem with this is that it always creates its own figure for it. and if I want to do anything further on it like select a line ROI I am limited to this way of plotting. What I want to achieve is the example below, but on a predefined figure, that I can update, so I do not need to create a new one every time I am plotting a signal. Am I missing something?

import hyperspy.api as hs
import scipy.misc
import matplotlib.pyplot as plt
im = hs.signals.Signal2D(scipy.misc.ascent())

line_roi = hs.roi.Line2DROI(0, 0, 512, 512, 10)

im.plot() #this is the part

roi1D = line_roi.interactive(im, color="yellow")

**Comment by jat255** [comment]
As far as I know, I don't think that's possible with `im.plot()`, but someone else probably knows the API internals better than I do.

I know the methods in the `hs.plot` API (such as `hs.plot.plot_images()`) has the ability to send the output to an existing figure, but I don't believe that works with interactivity, or at least it didn't work when I tried:

```python
%matplotlib ipympl 

import hyperspy.api as hs
import scipy.misc
import matplotlib.pyplot as plt
im = hs.signals.Signal2D(scipy.datasets.ascent())

f = plt.figure()

hs.plot.plot_images([im], fig=f)
```

The methods to look at would be [`BaseSignal.plot()`](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.plot), [`Signal2D.plot()`](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/Signal2D.html#hyperspy.api.signals.Signal2D.plot) and the like

**Reply by ericpre** [reply]
[`hs.plot.plot_spectra`](https://hyperspy.org/hyperspy-doc/current/reference/api.plot/index.html#hyperspy.api.plot.plot_spectra) does support some interactivity (but not all of the `plot` method); the figure should update when the data changes, as shown in this example: https://hyperspy.org/hyperspy-doc/current/auto_examples/region_of_interest/ExtractLineProfile.html#sphx-glr-auto-examples-region-of-interest-extractlineprofile-py

**Comment by ericpre** [comment, marked as answer]
Using an already defined matplotlib figure it not possible yet but it is being implemented as part of https://github.com/hyperspy/hyperspy/pull/3343.

What is your use case? Why do you need to put it in an existing matplotlib figure?

**Reply by Spooyler** [reply]
Thank you for the response. It is good to know it is being implemented.

I use hyperspy daily for analysis, but I find plotting with standard matplotlib more intuitive especially when I try to look at different type of data together.
